File: PPC/Assignments/KnapsackProblem2/KnapsackProblemKotlin/src/main/kotlin/knapsack/actor/actors/CrossoverActor.kt
package knapsack.actor.actors

import Individual
import KnapsackGA.Companion.tournament
import akka.actor.AbstractActor
import java.util.concurrent.ThreadLocalRandom

class CrossoverActor : AbstractActor() {
    data class Request(
        val population: Array<Individual>,
        val chunkSize: Int,
        val chunkIdx: Int
    )

    data class Response(val newChunk: Array<Individual>, val chunkIdx: Int)

    override fun createReceive(): Receive {
        return receiveBuilder()
            .match(Request::class.java) { request ->
                val population = request.population
                val chunkSize = request.chunkSize
                val newChunk = Array(chunkSize) { Individual() }

                val r = ThreadLocalRandom.current()
                for (i in 0 until chunkSize) {
                    val parent1 = tournament(r, population)
                    val parent2 = tournament(r, population)

                    newChunk[i] = parent1.crossoverWith(parent2, r)
                }

                sender.tell(Response(newChunk, request.chunkIdx), self)
            }
            .build()
    }
}

File: PPC/Assignments/KnapsackProblem2/KnapsackProblemGo/knapsack/actor/actors/crossoverActor.go
package actors

import (
	"KnapsackProblemGo/knapsack"
	"math/rand"
	"time"

	"github.com/asynkron/protoactor-go/actor"
)

type CrossoverActor struct{}

type CrossoverRequest struct {
	Population []*knapsack.Individual
	ChunkSize  int
	ChunkIdx   int
}

type CrossoverResponse struct {
	NewChunk []*knapsack.Individual
	ChunkIdx int
}

func NewCrossoverActor() actor.Actor {
	return &CrossoverActor{}
}

func (a *CrossoverActor) Receive(context actor.Context) {
	switch msg := context.Message().(type) {
	case *CrossoverRequest:
		a.handleRequest(msg, context)
	}
}

func (a *CrossoverActor) handleRequest(request *CrossoverRequest, context actor.Context) {
	population := request.Population
	chunkSize := request.ChunkSize
	idx := request.ChunkIdx
	newChunk := make([]*knapsack.Individual, chunkSize)

	r := rand.New(rand.NewSource(time.Now().UnixNano() + int64(idx)))

	for i := 0; i < chunkSize; i++ {
		parent1 := knapsack.Tournament(r, population)
		parent2 := knapsack.Tournament(r, population)

		newChunk[i] = parent1.CrossoverWith(parent2, r)
	}

	context.Respond(&CrossoverResponse{
		NewChunk: newChunk,
		ChunkIdx: idx,
	})
}
